### PR TITLE
#429 Task card component- redesign 

### LINKED ISSUE #429: Status Card Component
# Status card
## Overview

In Real Dev Squad, we use the [status site](https://status.realdevsquad.com/) to keep track of ongoing tasks. This a custom project tracking solution for our usecases.

It's time to refresh our designs and start making things a bit more consistent with the overall brand. So, we are starting off with updating or creating our components.

We want to implement this new design for cards that display each task.

## Design
[Figma Link](https://www.figma.com/file/NQco0iElImpT2ncuczuh2j/Status-Card?node-id=0%3A1&t=LrL6wSrQJYhw8Ypv-1)

## User stories

- As a viewer, the card should match the design provided.
- As a viewer, the card should correctly render the information to match the current implementation.
- As a superuser, the "edit" controls should be available as currently available.

## Requirements

- TDD should be followed.
- Put the component usage behind a FF(feature flag).

## Acceptance Criteria

- [ ] The new component can seemlessly be swapped out without affecting the functionality of the existing page.
- [ ] Tests coverage is exhaustive enough for the component to guarantee correctness and stability.

### PR DESCRIPTION
**Pull request by bhtibrewal**
closes #429 

[Figma Link](https://www.figma.com/file/NQco0iElImpT2ncuczuh2j/Status-Card?node-id=0%3A1&t=LrL6wSrQJYhw8Ypv-1)
Updating Task Card Component
### How to test the changes
the redesign is implemented behind a feature flag, to view it add `?dev=true` to the URL

**Packages Added**
Have added "[@testing-library/react-hooks": "^8.0.1](https://react-hooks-testing-library.com/usage/basic-hooks)" and "[@testing-library/dom": "^9.0.1](https://www.npmjs.com/package/@testing-library/dom)", packages as the functionality were required for testing react-hooks 

**Before**
image

**After**
image

-------
**Before**
image

**After**
image

### Video
https://user-images.githubusercontent.com/42600164/229362697-92e5c66f-16fa-4500-b0a8-9c497156717b.mov

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -9710,6 +9732,11 @@ lz-string@^1.4.4:
   resolved "https://registry.yarnpkg.com/lz-string/-/lz-string-1.4.4.tgz#c0d8eaf36059f705796e1e344811cf4c498d3a26"
   integrity sha512-0ckx7ZHRPqb0oUm8zNr+90mtf9DQB60H1wMCjBtfi62Kl3a7JbHob6gA2bC+xRvZoOL+1hzUK8jeuEIQE8svEQ==
 
+lz-string@^1.5.0:
+  version "1.5.0"
+  resolved "https://registry.yarnpkg.com/lz-string/-/lz-string-1.5.0.tgz#c1ab50f77887b712621201ba9fd4e3a6ed099941"
```
Comment (Maheima) [inline yarn.lock:9737]: do we need it here?

File: package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "@reduxjs/toolkit": "^1.9.1",
     "@testing-library/jest-dom": "^5.16.1",
     "@testing-library/react": "^13.3.0",
+    "@testing-library/react-hooks": "^8.0.1",
```
Comment (hjaintech) [inline package.json:32]: Should this be part of `devDependencies`?

File: src/components/tasks/card/index.tsx [hunk 3]
```diff
@@ -48,11 +49,14 @@ const Card: FC<Props> = ({
     const [showEditButton, setShowEditButton] = useState(false);
     const [keyLongPressed] = useKeyLongPressed();
     const [isLoading, setIsLoading] = useState<boolean>(false);
-
     const { actions, state } = useAppContext();
+    const router = useRouter();
+    const { query } = router;
+    const dev = !!query.dev;
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:55]: What does the variable `dev` do? If this is related to devMode, how about instead of using the terminology of dev vs prod, use `isNewCardEnabled` kind of terminology here?

File: src/components/tasks/card/index.tsx [hunk 4]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                {/* loading spinner */}
+                {isLoading && (
+                    <div className={classNames.loadingBg}>
+                        <div className={classNames.spinner}>
+                            <span className={classNames.screenReaderOnly}>
+                                loading
+                            </span>
+                        </div>
+                    </div>
+                )}
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:286]: Looks like this could be separate component

```suggestion
                {isLoading && <Loader />}
```

File: src/components/tasks/card/index.tsx [hunk 5]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                        className={classNames.cardTitle}
+                        contentEditable={shouldEdit}
+                        onKeyPress={(e) => handleChange(e, 'title')}
+                        role="button"
+                        tabIndex={0}
+                    >
+                        {cardDetails.title}
+                    </h2>
+                    {/* progress bar */}
+                    <div className={classNames.progressContainerUpdated}>
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:299]: Similarly, progress bar could be a separate component itself

File: src/components/tasks/card/index.tsx [hunk 6]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                    </div>
+                </div>
+                <div className={classNames.dateInfo}>
+                    <div>
+                        <span className={classNames.cardSpecialFont}>
+                            Estimated completion{' '}
+                        </span>
+                        {renderDate(fromNowEndsOn, shouldEdit)}
+                    </div>
+                    <span
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:325]: See if you can break down this JSX into multiple meaningful presentational components. Right now it is one big chunk of JSX which is hard to read and debug.

File: src/components/tasks/card/index.tsx [hunk 7]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                        className={classNames.cardSpecialFont}
+                        contentEditable={shouldEdit}
+                        onKeyPress={(e) => handleChange(e, 'startedOn')}
+                        role="button"
+                        tabIndex={0}
+                    >
+                        {cardDetails.status == 'Available'
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:332]: Always prefer `===`
```suggestion
                        {cardDetails.status === 'Available'
```

Comment (hjaintech) [inline src/components/tasks/card/index.tsx:332]: For fixed values like possible values of status, it is better to create a enum/constant and use it here. Removes the mistakes of typos as well.

File: src/components/tasks/card/index.tsx [hunk 8]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                            ? 'Not started '
+                            : `Started on ${fromNowStartedOn}`}
+                    </span>
+                </div>
+
+                <div className={classNames.cardItems}>
+                    <div className={classNames.contributor}>
+                        <span className={classNames.cardSpecialFont}>
+                            Assigned to{' '}
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:341]: Don't put a fixed space like this. Use CSS instead.

File: src/components/tasks/card/index.tsx [hunk 9]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                        </span>
+                        <span className={classNames.contributorImage}>
+                            <Image
+                                src={assigneeProfilePic}
+                                alt="Assignee profile picture"
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:346]: This would be a better alt text
```suggestion
                                alt="{`profile picture of ${{cardDetails.assignee}}`}"
```

File: src/components/tasks/card/index.tsx [hunk 10]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                                onError={contributorImageOnError}
+                                width={30}
+                                height={30}
+                            />
+                        </span>
+                        <span
+                            className={classNames.cardStrongFont}
+                            contentEditable={shouldEdit}
+                            onKeyPress={(e) => handleChange(e, 'assignee')}
+                            role="button"
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:356]: Why is this element having the role as button?

File: src/components/tasks/card/index.tsx [hunk 11]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                </div>
+
+                {isUserAuthorized && showEditButton && (
+                    <div
+                        className={classNames.editButton}
+                        data-testid="edit-button"
+                    >
+                        <Image
+                            src="/pencil.webp"
+                            alt="edit Pencil"
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:413]:
```suggestion
                            alt="edit"
```

File: src/components/tasks/card/index.tsx [hunk 12]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                            width={iconWidth}
+                            height={iconHeight}
+                            onClick={onEditEnabled}
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:416]: See if this is reachable via keyboard. If not, then should add tabIndex=0 here probably.

File: src/components/tasks/card/index.tsx [hunk 13]
```diff
@@ -258,13 +262,171 @@ const Card: FC<Props> = ({
...
+                            shouldEdit && classNames.editMode
+                        }`}
+                    >
+                        <div className={classNames.taskTagLevelContainer}>
+                            {taskTagLevel?.map((item) => (
+                                <span
+                                    key={item.tagId}
+                                    className={classNames.taskTagLevel}
+                                >
+                                    {item.tagName}{' '}
```
Comment (hjaintech) [inline src/components/tasks/card/index.tsx:373]: Again, don't add static empty string. Use CSS for spacing

Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: done

File: src/components/tasks/card/index.tsx (hunk 11 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: done

File: src/components/tasks/card/index.tsx (hunk 10 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:368]: is was mentioned in the old code, so I didn't change it

File: src/components/tasks/card/index.tsx (hunk 8 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: done

File: src/components/tasks/card/index.tsx (hunk 7 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: done

Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: done

File: src/components/tasks/card/index.tsx (hunk 12 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: yes checked it was not added tabIndex

File: src/components/tasks/card/index.tsx (hunk 3 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx]: changed it

File: src/components/tasks/card/index.tsx (hunk 5 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:326]: done

File: src/components/tasks/card/Loader.tsx [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import classNames from '@/components/tasks/card/card.module.scss';
```
Comment (hjaintech) [inline src/components/tasks/card/Loader.tsx:1]: We should add unit tests for these components as well.

File: src/components/tasks/card/index.tsx [hunk 15]
```diff
@@ -257,25 +263,148 @@ const Card: FC<Props> = ({
...
+                        {renderDate(fromNowEndsOn, shouldEdit)}
+                    </div>
+                    <span
+                        className={classNames.cardSpecialFont}
+                        contentEditable={shouldEdit}
+                        onKeyPress={(e) => handleChange(e, 'startedOn')}
+                        role="button"
+                        tabIndex={0}
+                    >
+                        {cardDetails.status === AVAILABLE
```
[the remark this answers is not in the document]
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:345]: done changed it please see interfaces/task-status.ts

File: src/components/tasks/card/TaskLevelMap.tsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
+import classNames from '@/components/tasks/card/card.module.scss';
+import { isUserAuthorizedContext } from '@/context/isUserAuthorized';
+import taskItem from '@/interfaces/taskItem.type';
+import { useContext } from 'react';
+type Props = {
+    taskTagLevel: taskItem[] | undefined;
```
Comment (vvaibhavdesai) [inline src/components/tasks/card/TaskLevelMap.tsx:6]: use `taskTagLevel?: taskItem[]`

File: src/components/tasks/card/index.tsx [hunk 17]
```diff
@@ -257,25 +265,148 @@ const Card: FC<Props> = ({
...
+            onKeyPress={(e) => handleChange(e, 'title')}
+            role="button"
+            tabIndex={0}
+        >
+            {cardDetails.title}
+        </h2>
+    );
+
+    // show redesign only on dev
+    if (isNewCardEnabled)
```
Comment (vvaibhavdesai) [inline src/components/tasks/card/index.tsx:311]: this piece of code needs to be refactored

**Review by vvaibhavdesai — COMMENTED** [review]
Great work 🚀 
I see `isAuthorized `and `showEditButton  `and another `prop `being used for `&&` check. which seems a lot repetitive is there any way to make it into custom hooks ?

File: src/components/tasks/card/index.tsx (hunk 17 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:328]: will be tracking this with Issue #497

File: src/components/tasks/card/Loader.tsx (hunk 14 above)
Comment (bhtibrewal) [reply src/components/tasks/card/Loader.tsx]: done

File: src/components/tasks/card/index.tsx [hunk 18]
```diff
@@ -257,25 +263,148 @@ const Card: FC<Props> = ({
...
+        return (
```
[the remark this answers is not in the document]
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:329]: will be tracking this with Issue #497

File: src/components/tasks/card/index.tsx [hunk 19]
```diff
@@ -257,25 +263,148 @@ const Card: FC<Props> = ({
...
+                alt="edit"
+                width={iconWidth}
+                height={iconHeight}
+                onClick={onEditEnabled}
+                tabIndex={0}
+            />
+        </div>
+    );
+
+    const ProgressIndicator = () => (
```
[the remark this answers is not in the document]
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:298]: will be tracking this with Issue #497

File: src/components/tasks/card/index.tsx (hunk 6 above)
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:357]: done partially, will be tracking this with Issue #497

File: src/components/tasks/card/TaskLevelMap.tsx (hunk 16 above)
Comment (bhtibrewal) [reply src/components/tasks/card/TaskLevelMap.tsx]: done

File: src/components/tasks/card/index.tsx [hunk 20]
```diff
@@ -257,25 +263,148 @@ const Card: FC<Props> = ({
...
+                {/* loading spinner */}
+                {isLoading && <Loader />}
+
+                <div className={classNames.cardItems}>
+                    <CardTitle />
+                    {/* progress bar */}
+                    <div className={classNames.progressContainerUpdated}>
+                        <ProgressIndicator />
+                        <span>{content.percentCompleted}% </span>
+                    </div>
```
[the remark this answers is not in the document]
Comment (bhtibrewal) [reply src/components/tasks/card/index.tsx:348]: will be tracking this with Issue https://github.com/Real-Dev-Squad/website-status/issues/497

File: package.json [hunk 21]
```diff
@@ -1,104 +1,104 @@
...
-    "node": "18.14.2",
-    "yarn": "1.22.19"
-  }
+    "name": "status",
+    "version": "1.0.0",
+    "description": "Shows a status of the ongoing projects being done",
+    "main": "index.js",
+    "scripts": {
+        "run-command:dev-server": "next dev -p 3000",
+        "run-command:reverse-ssl": "local-ssl-proxy --source 443 --target 3000",
```
Comment (Maheima) [inline package.json:8]: can you please revert these changes?

**Review by Maheima — CHANGES_REQUESTED** [review]

File: package.json (hunk 21 above)
Comment (bhtibrewal) [reply package.json]: done

**Review by Maheima — APPROVED** [review]
LGTM!

**Review by vvaibhavdesai — APPROVED** [review]
